File: Code/Erlang/chaos_monkey.erl

% This module encompasses the "chaos monkey" program used to randomly terminate/kill processes
% 
% Author: Andreea Lutac

-module(chaos_monkey).
-author('andreea.lutac').
-import(lists, [nth/2, seq/2]).
-export([start/1, kill_sup/1]).


start(Args) ->
	KillN = nth(1, Args),
	WorkerN = nth(2, Args),
	KillType = nth(3, Args),
	io:format("Starting to kill random processes....~n",[]),
	kill(KillN, WorkerN, KillType).


kill(0, WorkerN, KillType) ->
	io:format("Finished killing processes.~n",[]);
	
kill(N, WorkerN, KillType) ->
	Proc = [ global:whereis_name(I) || I <- seq(1, WorkerN)],
	FilterProc = [ P || P <- Proc, P/=undefined ],
	Rand = trunc(rand:uniform()*length(FilterProc)+1),

	case KillType of
		stop ->
			nth(Rand,FilterProc) ! KillType;
		kill ->
			exit(nth(Rand,FilterProc),kill)
	end,

	kill(N-1, WorkerN, KillType).

kill_sup(Type) ->
	if
		Type==top ->
			gen_server:cast(whereis(top_supervisor), stop);
		Type==face ->
			gen_server:cast(whereis(face_supervisor), stop);
		Type==agg ->
			gen_server:cast(whereis(aggregator_supervisor), stop);
		true ->
			io:format("Unkown type. ~n",[])
	end,
	
	timer:sleep(5),
	io:format("Killed supervisor. ~n", []).	

File: Code/ROS/chaos_monkey.py
## This script randomly terminates a number of ROS nodes,
## either "gracefully" or brutally.
##
## Author: Andreea Lutac

from subprocess import call
import sys
import random

kill_n = sys.argv[1]
worker_n = sys.argv[2]
kill_type = sys.argv[3]

for x in random.sample(range(0, int(worker_n)), int(kill_n)):
	if kill_type == "stop":
		call(["rosnode", "kill","/orbit_face_tracking_n"+str(x)])
	elif kill_type == "kill":
		call(["pkill", "-f", "orbit_face_tracking_n"+str(x),"-9"])
	elif kill_type == "main":
		call(["pkill", "-f", "orbit_face_tracking_main","-9"])

print "Finished killing."